flink-table-store-core/src/main/java/org/apache/flink/table/store/file/io/SingleFileWriter.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.table.store.file.io;
+
+import org.apache.flink.api.common.serialization.BulkWriter;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.store.file.utils.FileUtils;
+import org.apache.flink.util.IOUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.function.Function;
+
+/**
+ * A {@link FileWriter} to produce a single file.
+ *
+ * @param <T> type of records to write.
+ * @param <R> type of result to produce after writing a file.
+ */
+public abstract class SingleFileWriter<T, R> implements FileWriter<T, R> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SingleFileWriter.class);
+
+    private final BulkWriter.Factory<RowData> factory;
+    protected final Path path;
+    private final Function<T, RowData> converter;
+
+    private long recordCount;
+    private long length;
+    protected boolean closed;
+
+    private BulkWriter<RowData> writer;
+    private FSDataOutputStream out;
+
+    public SingleFileWriter(
+            BulkWriter.Factory<RowData> factory, Path path, Function<T, RowData> converter) {
+        this.factory = factory;
+        this.path = path;
+        this.converter = converter;
+
+        this.recordCount = 0;
+        this.length = 0;
+        this.closed = false;
+
+        this.writer = null;
+        this.out = null;
+    }
+
+    public Path path() {
+        return path;
+    }
+
+    @Override
+    public void write(T record) throws IOException {
+        writeImpl(record);
+    }
+
+    protected RowData writeImpl(T record) throws IOException {
+        if (closed) {
+            throw new RuntimeException("Writer has already closed!");
+        }
+
+        try {
+            if (writer == null) {
```

Comment:
It is a single file writer, the file should be single instead of zero.
We can create writer in the constructor. This can avoid various inconsistencies caused by not producing files.